File: system-test/dotnet/SystemTests/Legacy/Mod04/Base/BaseClientTest.cs
using SystemTests.TestInfrastructure.Configuration;
using Dsl.Core;
using Driver.Adapter.External.Erp.Client;
using Driver.Adapter.Shop.Api.Client;
using Driver.Adapter.Shop.Ui.Client;
using Driver.Adapter.External.Tax.Client;
using Xunit;

namespace SystemTests.Legacy.Mod04.Base;

public abstract class BaseClientTest : BaseConfigurableTest, IAsyncLifetime
{
    protected readonly Dsl.Core.Configuration _configuration;

    protected ShopUiClient? _shopUiClient;
    protected ShopApiClient? _shopApiClient;

    protected ErpRealClient? _erpClient;
    protected TaxRealClient? _taxClient;

    protected BaseClientTest()
    {
        _configuration = LoadConfiguration();
    }

    public virtual Task InitializeAsync()
    {
        return Task.CompletedTask;
    }

    protected async Task SetUpShopUiClientAsync()
    {
        _shopUiClient = await ShopUiClient.CreateAsync(_configuration.ShopUiBaseUrl);
    }

    protected void SetUpShopApiClient()
    {
        _shopApiClient = new ShopApiClient(_configuration.ShopApiBaseUrl);
    }

    protected void SetUpExternalClients()
    {
        _erpClient = new ErpRealClient(_configuration.ErpBaseUrl);
        _taxClient = new TaxRealClient(_configuration.TaxBaseUrl);
    }

    public virtual async Task DisposeAsync()
    {
        if (_shopUiClient != null)
            await _shopUiClient.DisposeAsync();
        _shopApiClient?.Dispose();
        _erpClient?.Dispose();
        _taxClient?.Dispose();
    }
}

File: system-test/java/src/test/java/com/optivem/shop/systemtest/legacy/mod04/base/BaseClientTest.java
package com.optivem.shop.systemtest.legacy.mod04.base;

import com.optivem.shop.systemtest.configuration.BaseConfigurableTest;
import com.optivem.shop.systemtest.configuration.Configuration;
import com.optivem.shop.dsl.driver.adapter.external.erp.client.ErpRealClient;
import com.optivem.shop.dsl.driver.adapter.external.tax.client.TaxRealClient;
import com.optivem.shop.dsl.driver.adapter.shop.api.client.ShopApiClient;
import com.optivem.shop.dsl.driver.adapter.shop.ui.client.ShopUiClient;
import com.optivem.shop.systemtest.infrastructure.playwright.BrowserLifecycleExtension;
import com.optivem.shop.dsl.common.Closer;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;

public class BaseClientTest extends BaseConfigurableTest {
    protected Configuration configuration;

    protected ShopUiClient shopUiClient;
    protected ShopApiClient shopApiClient;
    protected ErpRealClient erpClient;
    protected TaxRealClient taxClient;

    @BeforeEach
    protected void setUpConfiguration() {
        configuration = loadConfiguration();
    }

    protected void setUpShopUiClient() {
        shopUiClient = new ShopUiClient(configuration.getShopUiBaseUrl(), BrowserLifecycleExtension.getBrowser());
    }

    protected void setUpShopApiClient() {
        shopApiClient = new ShopApiClient(configuration.getShopApiBaseUrl());
    }

    protected void setUpExternalClients() {
        erpClient = new ErpRealClient(configuration.getErpBaseUrl());
        taxClient = new TaxRealClient(configuration.getTaxBaseUrl());
    }

    @AfterEach
    void tearDown() {
        Closer.close(shopUiClient);
        Closer.close(shopApiClient);
        Closer.close(erpClient);
        Closer.close(taxClient);
    }

}
